paastech-cloud/website
# API calls and User Authentication

## What does this PR do ?
- Adds autogenerated *api/generated* folder with functions for all API requests based on the OpenAPI spec;
> For now, the app uses Bearer token in header to authenticate user, I will remove it as soon as Fabien's pull request for cookie authentication is merged to main.
> You can use api objects in *api/api.ts* to make API requests for each respective controller. For example:
```typescript
    authApi.authControllerLogin(values)
    .then((response) => {
              if (response?.data?.accessToken) {
                  setAccessToken(response.data);
                  ...
              }
    })
    .catch((err) => {
        ...
    });
```
> or
```typescript
const { isLoading, isSuccess } = useQuery('verify email', () =>
      authApi.authControllerConfirmEmail(token));
```
> When you add endpoints/controllers to the API, make sure to grab the newest version of the OpenAPI spec and put it into *openapi.json* in the root of the website project. Then execute `npm run openapi` to update the autogenerated functions.
> If new controllers were added, you will need to add objects for them in the *api/api.ts* file similarly to already existing ones.

- Calls API to register user, verify their email, log them in and show their profile.
> I have added a custom hook `useUser` to help easily retrieve user object whenever needed (in profile page and in the navigation bar).
> To verify quickly if a user is authenticated without sending requests to the API and waiting for response, I have added a function `accessExpired`. It depends on the time of access expiration I add in the sessionStorage on login. It is not ideal since we'll have to make sure to keep the expiration time same between web and API (for now, it's 6 hours).

### How should this be manually tested?
- Step 1 : `npm run dev`
- Step 2 : register user
- Step 3 :  validate email (notice!: apparently the links now need to have a hash at the start to work, example: *http://localhost:5173/#/register*)
- Step 4: log the user in
  ...

### Other changes
The app depends on the route `GET /users` that I'll be adding to the API shortly. It is obviously necessary for a user to be able to get their profile without indicating what their username is. I am also enabling cors in the API.
